mojo/proposals/drain-collection-support.md

```diff
@@ -0,0 +1,452 @@
+tuples supported move unpacking. Reference issue
+[#5330](https://github.com/modular/modular/issues/5330).
+
+The basic case is a complete drain of each respective collection.
+
+```mojo
+def basic_user_experience_cases():
+    print('Testing list')
+    var l = List[Foo](Foo(x="a"), Foo(x="b"), Foo(x="c"))
+    for ref item in l.take_items():
+        take_ownership(item.take())
+    assert_equal(len(l), 0)
+
+    print('Testing set')
+    var set: Set[Foo] = {Foo(x="a"), Foo(x="b"), Foo(x="c")}
+    for ref item in set.take_items():
+        take_ownership(item.take())
+    assert_equal(len(set), 0)
+
+    print('Testing dict')
+    var dict: Dict[String, Foo] = {
+        "a": Foo(x="a"), 
+        "b": Foo(x="b"), 
+        "c": Foo(x="c")
+    }
+    for ref item in dict.take_items():
+        # Dict take_items returns the DictEntry struct, which does not 
+        # automatically unpack itself.
+        # Ideally mojo would support `__unpack__` like in #5330
+        # However we can define method that returns a 
+        # Tuple[Optional[Key], Optional[Value]]
+        # using optional allows for moving the key and value out of the item 
+        # without deinitializing the item overall.
+        var key_value = item.reap()
+        take_ownership(key_value[0].take(), key_value[1].take())
+    assert_equal(len(dict), 0)
+```
+
+Draining only slices of the collections. This moves the elements out of the
+List, then allows the list to do a single move-left on initialization instead of
+per element.
+
+```mojo
+def slice_drain_user_experience_cases():
+    print('Testing list')
+    var l = List[Foo](Foo(x="a"), Foo(x="b"), Foo(x="c"))
+    for ref item in l.take_items(Slice(1, 2)):
+        take_ownership(item.take())
+    assert_equal(len(l), 1)
+```
+
+Draining multiple lists via zip.
+
+```mojo
+    var keys = List[String]("a", "b", "c")
+    var values = List[Foo](Foo(x="d"), Foo(x="e"), Foo(x="f"))
+    var new_dict = Dict[String, Foo]()
+    for ref item in zip(keys, values):
+        new_dict[item[0].take()] = item[1].take()
+```
+
+Clearing collections on incomplete iteration.
+
+```mojo
+    print('Testing list')
+    var l = List[Foo](Foo(x="a"), Foo(x="b"), Foo(x="c"))
+    for ref item in l.take_items():
+        break # Iterator clears the remaining elements on destruction
+    assert_equal(len(l), 0)
+
+    print('Testing set')
+    var set: Set[Foo] = {Foo(x="a"), Foo(x="b"), Foo(x="c")}
+    for ref item in set.take_items():
+        break # Iterator clears the remaining elements on destruction
+    assert_equal(len(set), 0)
+
+    print('Testing dict')
+    var dict: Dict[String, Foo] = {
+        "a": Foo(x="a"), 
+        "b": Foo(x="b"), 
+        "c": Foo(x="c")
+    }
+    for ref item in dict.take_items():
+      break # Iterator clears the remaining elements on destruction
+    assert_equal(len(dict), 0)
+```
+
+## Questions
+
+**Do we think Mojo's Trait API is flexible enough to make Drain/TakeIterator a
+trait?**
```

Comment:
I don't think there is much value in making Drain/TakeIterator a trait. 
The only traits you should need to work with are `Iterator` and `Iterable`. `take_items` is just a way to get a specific kind of iterator you can then pass to generic functions.
```
fn foo(iter: Some[Iterable]):
    ...

var list = [1, 2, 3]
foo(list.take_items(range(1, 2)))
```